src/Pos/Explorer/Web/Server.hs

```diff
@@ -250,6 +304,45 @@ getTxSummary cTxId = do
 -- Helpers
 --------------------------------------------------------------------------------
 
+-- | Find all `MainBlock` by applying the *predicate*, starting from *headerHash*
+filterMainBlocks :: (SscHelpersClass ssc, ExplorerMode m) =>
+    HeaderHash ->
+    (MainBlock ssc -> Bool) ->
+    m [MainBlock ssc]
+filterMainBlocks headerHash predicate = rights <$> generalBlockSearch
+  where
+    generalBlockSearch = filterAllBlocks headerHash specializedPred (pure [])
+    specializedPred block = case block of
+        Left  _   -> False
+        Right mb  -> predicate mb
+
+-- | Find all blocks matching the sent predicate. This is a generic function
+-- that can be called with either `MainBlock` or `GenesisBlock` in mind.
+filterAllBlocks :: (SscHelpersClass ssc, ExplorerMode m) =>
+    HeaderHash ->
+    (Block ssc -> Bool) ->
+    m [Block ssc] ->
+    m [Block ssc]
+filterAllBlocks headerHash predicate acc
+    -- When we reach the genesis block, return the accumulator.
+    | headerHash == genesisHash = acc
+    -- Otherwise iterate back from the top block (called tip) and add all
+    -- blocks (hash) to accumulator satisfying the predicate.
+    | otherwise = do
+        -- Get the block with the sent hash, throw exception if/when the block
+        -- search fails.
+        block <- DB.getBlock headerHash >>=
+            maybeThrow (Internal "Block with hash cannot be found!")
+        -- If there is a block then iterate backwards with the predicate
+        let prevBlock = block ^. prevBlockL
+
+        case predicate block of
```

Comment:
Fixed. I think if should be avoided in favor of case matching which should be idiomatic Haskell, right?